app/controllers/users/nbp_wallet_controller.rb

```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+require 'rqrcode'
+
+module Users
+  class NbpWalletController < ApplicationController
+    skip_before_action :require_user!
+    skip_after_action :verify_authorized
+    before_action :set_provider_details
+    before_action :set_connector_api
+
+    def connect
+      if session[:relationship_template_id].present?
+        @connector.reject_all_relationships(session[:relationship_template_id])
+      end
+      template_id, truncated_reference = @connector.create_relationship_template
+      session[:relationship_template_id] = template_id
+      session[:relationship_template_tr] = truncated_reference
+
+      @template_url = "nmshd://tr##{truncated_reference}"
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      redirect_to new_user_registration_path, alert: t('common.errors.generic_try_later')
+    end
+
+    def qr_code
+      send_data RQRCode::QRCode.new("nmshd://tr##{session[:relationship_template_tr]}").as_png(border_modules: 0), type: 'image/png'
+    end
+
+    def relationship_status
+      render json: {status: 'missing relationship_template_id'} and return if session[:relationship_template_id].nil?
+
+      relationships = @connector.relationships(session[:relationship_template_id], 'Pending')
+
+      render json: {status: 'waiting'} and return if relationships.empty?
+
+      render json: {status: 'ready'}
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      redirect_to nbp_wallet_connect_users_path, alert: t('common.errors.generic')
+    end
+
+    def finalize
+      relationships = @connector.relationships(session[:relationship_template_id], 'Pending')
+      abort_and_refresh and return if relationships.size != 1
+
+      accept_and_create_user(relationships.first)
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      abort_and_refresh
+    end
+
+    private
+
+    def accept_and_create_user(relationship) # rubocop:disable Metrics/AbcSize
+      relationship_changes = relationship[:changes]
+      abort_and_refresh if relationship_changes.size != 1
+
+      user_attributes = @connector.parse_userdata_in_rel_change(relationship_changes.first)
+
+      @user = User.new_from_omniauth('nbp', @provider_uid, user_attributes)
+      return abort_and_refresh("Could not create User: #{@user.errors.full_messages.join(', ')}") unless @user.save
+
+      @user.identities << UserIdentity.new(omniauth_provider: 'enmeshed', provider_uid: relationship[:peer])
+
+      if @connector.accept_relationship_change(relationship[:id], relationship_changes.first[:id])
+        @user.send_confirmation_instructions
+        flash[:notice] = t('devise.registrations.signed_up_but_unconfirmed')
+        redirect_to home_index_path
```

Comment:
I managed to find an edge case, where I got the confirmation instructions but still was redirected to the `nbp_wallet_connect_users_path`. Unfortunately, I missed the flash message (due to the mail opening) and thus don't know what went wrong. Do you have an idea?

<details><summary>Details</summary>
<p>

```
Started POST "/users/nbp_wallet/relationship_status" for 127.0.0.1 at 2024-05-12 23:30:55 +0300
Processing by Users::NbpWalletController#relationship_status as */*
Completed 200 OK in 40ms (Views: 0.1ms | ActiveRecord: 0.0ms | Allocations: 3151)


Started GET "/users/nbp_wallet/finalize" for 127.0.0.1 at 2024-05-12 23:30:55 +0300
Processing by Users::NbpWalletController#finalize as HTML
  TRANSACTION (0.8ms)  BEGIN
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
  User Exists? (24.5ms)  SELECT 1 AS one FROM "users" WHERE "users"."email" = $1 LIMIT $2  [["email", "p14@example.org"], ["LIMIT", 1]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
  User Exists? (1.0ms)  SELECT 1 AS one FROM "users" WHERE LOWER("users"."email") = LOWER($1) LIMIT $2  [["email", "p14@example.org"], ["LIMIT", 1]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
  UserIdentity Exists? (1.2ms)  SELECT 1 AS one FROM "user_identities" WHERE "user_identities"."provider_uid" = $1 AND "user_identities"."omniauth_provider" = $2 LIMIT $3  [["provider_uid", "p14@example.com"], ["omniauth_provider", "nbp"], ["LIMIT", 1]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
  User Create (5.0ms)  INSERT INTO "users" ("first_name", "last_name", "created_at", "updated_at", "role", "deleted", "email", "encrypted_password", "reset_password_token", "reset_password_sent_at", "remember_created_at", "confirmation_token", "confirmed_at", "confirmation_sent_at", "unconfirmed_email", "failed_attempts", "unlock_token", "locked_at", "preferred_locale", "password_set") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20) RETURNING "id"  [["first_name", "Max"], ["last_name", "Mustermann 14"], ["created_at", "2024-05-12 20:30:55.803827"], ["updated_at", "2024-05-12 20:30:55.803827"], ["role", "user"], ["deleted", nil], ["email", "p14@example.org"], ["encrypted_password", "[FILTERED]"], ["reset_password_token", "[FILTERED]"], ["reset_password_sent_at", "[FILTERED]"], ["remember_created_at", nil], ["confirmation_token", "[FILTERED]"], ["confirmed_at", "2024-05-12 20:30:55.763044"], ["confirmation_sent_at", nil], ["unconfirmed_email", nil], ["failed_attempts", 0], ["unlock_token", "[FILTERED]"], ["locked_at", nil], ["preferred_locale", nil], ["password_set", "[FILTERED]"]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
  UserIdentity Exists? (3.6ms)  SELECT 1 AS one FROM "user_identities" WHERE "user_identities"."provider_uid" = $1 AND "user_identities"."omniauth_provider" = $2 LIMIT $3  [["provider_uid", "p14@example.com"], ["omniauth_provider", "nbp"], ["LIMIT", 1]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
Started POST "/users/nbp_wallet/relationship_status" for 127.0.0.1 at 2024-05-12 23:30:56 +0300
Processing by Users::NbpWalletController#relationship_status as */*
Completed 200 OK in 38ms (Views: 0.1ms | ActiveRecord: 0.0ms | Allocations: 3112)


Started GET "/users/nbp_wallet/finalize" for 127.0.0.1 at 2024-05-12 23:30:56 +0300
Processing by Users::NbpWalletController#finalize as HTML
  UserIdentity Create (621.0ms)  INSERT INTO "user_identities" ("user_id", "omniauth_provider", "provider_uid", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["user_id", 83], ["omniauth_provider", "nbp"], ["provider_uid", "p14@example.com"], ["created_at", "2024-05-12 20:30:55.831101"], ["updated_at", "2024-05-12 20:30:55.831101"]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
  TRANSACTION (1.6ms)  COMMIT
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
  TRANSACTION (1.2ms)  BEGIN
  ↳ app/controllers/users/nbp_wallet_controller.rb:64:in `accept_and_create_user'
  UserIdentity Exists? (2.8ms)  SELECT 1 AS one FROM "user_identities" WHERE "user_identities"."provider_uid" = $1 AND "user_identities"."omniauth_provider" = $2 LIMIT $3  [["provider_uid", "id1Mky2Aoyky2pKd5CJmsvQW4zGJcfK1KVSn"], ["omniauth_provider", "enmeshed"], ["LIMIT", 1]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:64:in `accept_and_create_user'
  UserIdentity Create (0.9ms)  INSERT INTO "user_identities" ("user_id", "omniauth_provider", "provider_uid", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["user_id", 83], ["omniauth_provider", "enmeshed"], ["provider_uid", "id1Mky2Aoyky2pKd5CJmsvQW4zGJcfK1KVSn"], ["created_at", "2024-05-12 20:30:56.481804"], ["updated_at", "2024-05-12 20:30:56.481804"]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:64:in `accept_and_create_user'
  TRANSACTION (0.6ms)  COMMIT
  ↳ app/controllers/users/nbp_wallet_controller.rb:64:in `accept_and_create_user'
  TRANSACTION (0.8ms)  BEGIN
  ↳ app/controllers/users/nbp_wallet_controller.rb:67:in `accept_and_create_user'
  User Update (4.7ms)  UPDATE "users" SET "updated_at" = $1, "confirmation_token" = $2, "confirmation_sent_at" = $3 WHERE "users"."id" = $4  [["updated_at", "2024-05-12 20:30:56.705119"], ["confirmation_token", "[FILTERED]"], ["confirmation_sent_at", "2024-05-12 20:30:56.702097"], ["id", 83]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:67:in `accept_and_create_user'
  TRANSACTION (0.6ms)  COMMIT
  ↳ app/controllers/users/nbp_wallet_controller.rb:67:in `accept_and_create_user'
  Rendering users/mailer/confirmation_instructions.html.slim
  Rendered users/mailer/confirmation_instructions.html.slim (Duration: 3.4ms | Allocations: 3135)
  Rendering users/mailer/confirmation_instructions.text.slim
  Rendered users/mailer/confirmation_instructions.text.slim (Duration: 2.3ms | Allocations: 2583)
Users::DeviseMailer#confirmation_instructions: processed outbound mail in 41.0ms
  TRANSACTION (1.3ms)  BEGIN
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
  User Exists? (60.9ms)  SELECT 1 AS one FROM "users" WHERE "users"."email" = $1 LIMIT $2  [["email", "p14@example.org"], ["LIMIT", 1]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
Delivered mail 66412700ba0d3_dd5efec-31e@Sebastians-MacBook-Pro.local.mail (60.3ms)

Redirected to http://localhost:7500/home
Completed 302 Found in 1425ms (ActiveRecord: 682.0ms | Allocations: 207877)


  User Exists? (2.9ms)  SELECT 1 AS one FROM "users" WHERE LOWER("users"."email") = LOWER($1) LIMIT $2  [["email", "p14@example.org"], ["LIMIT", 1]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
  UserIdentity Exists? (3.8ms)  SELECT 1 AS one FROM "user_identities" WHERE "user_identities"."provider_uid" = $1 AND "user_identities"."omniauth_provider" = $2 LIMIT $3  [["provider_uid", "p14@example.com"], ["omniauth_provider", "nbp"], ["LIMIT", 1]]
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
  TRANSACTION (0.6ms)  ROLLBACK
  ↳ app/controllers/users/nbp_wallet_controller.rb:62:in `accept_and_create_user'
  TRANSACTION (0.7ms)  BEGIN
  ↳ app/models/user.rb:65:in `call'
  TRANSACTION (0.6ms)  COMMIT
  ↳ app/models/user.rb:65:in `call'
Redirected to http://localhost:7500/users/nbp_wallet/connect
Completed 302 Found in 565ms (ActiveRecord: 157.9ms | Allocations: 155843)


Started GET "/users/nbp_wallet/connect" for 127.0.0.1 at 2024-05-12 23:30:56 +0300
Processing by Users::NbpWalletController#connect as HTML
  Rendering layout layouts/application.html.slim
  Rendering users/nbp_wallet/connect.html.slim within layouts/application
  Rendered users/nbp_wallet/connect.html.slim within layouts/application (Duration: 0.6ms | Allocations: 240)
  Rendered application/_breadcrumbs_and_title.html.slim (Duration: 1.5ms | Allocations: 1915)
  Rendered application/_navigation.html.slim (Duration: 0.1ms | Allocations: 73)
  Rendered application/_locale_selector.html.slim (Duration: 0.2ms | Allocations: 242)
  Rendered application/_navigation_collection_link.html.slim (Duration: 0.1ms | Allocations: 117)
  Rendered application/_navigation_collection_link.html.slim (Duration: 0.0ms | Allocations: 118)
  Rendered application/_session.html.slim (Duration: 0.2ms | Allocations: 606)
  Rendered application/_flash.html.slim (Duration: 0.8ms | Allocations: 113)
  Rendered application/_footer.html.slim (Duration: 0.1ms | Allocations: 237)
  Rendered layout layouts/application.html.slim (Duration: 16.9ms | Allocations: 20216)
Completed 200 OK in 265ms (Views: 18.2ms | ActiveRecord: 0.0ms | Allocations: 30145)


Started GET "/users/nbp_wallet/qr_code" for 127.0.0.1 at 2024-05-12 23:30:57 +0300
Processing by Users::NbpWalletController#qr_code as HTML
  Rendering text template
  Rendered text template (Duration: 0.1ms | Allocations: 26)
Sent data  (8.1ms)
Completed 200 OK in 48ms (Views: 8.1ms | ActiveRecord: 0.0ms | Allocations: 86546)


Started POST "/users/nbp_wallet/relationship_status" for 127.0.0.1 at 2024-05-12 23:30:58 +0300
Processing by Users::NbpWalletController#relationship_status as */*
Completed 200 OK in 46ms (Views: 0.2ms | ActiveRecord: 0.0ms | Allocations: 2776)
```

</p>
</details>

Reply:
One way I could reproduce a similar issue was to reuse an existing email address in the mobile app. However, for this occasion, no confirmation email was sent. The root cause of the previous issue is still unknown to me.

Reply:
> I managed to find an edge case, where I got the confirmation instructions but still was redirected to the `nbp_wallet_connect_users_path`. Unfortunately, I missed the flash message (due to the mail opening) and thus don't know what went wrong. Do you have an idea?
> Details

Thats weird. Looking at the code path I think the only way this could happen is if ```@user.send_confirmation_instructions``` threw a ```Faraday::ConnectionFailed``` because then the error handling in the ```finalize``` method would catch it and redirect to ```nbp_wallet_connect_users_path```. Other than that it should be impossible to get redirected to ```nbp_wallet_connect_users_path``` after the email was sent :thinking:

Reply:
I changed the frontend to only send one request to #finalize. I tested it with a sleep(1) in the #finalize action before and after the change and it looks like its fixed.

Reply:
Thanks!

Just for reference: We tracked the issue down to an increased latency during my previous test, with the `finalize` method being called twice simultaneously. This happened since the call to `relationship_status` was responding with `ready` (as the contact request was accepted) and thus the frontend was calling `finalize`. To finalize the account sign up as quickly as possible, we called `relationship_status` each second, causing this issue. With the new workaround, we ensure that `finalize` is only called once.